config/pmd.xml

```diff
@@ -7,6 +7,35 @@
   <description>
     PMD common ruleset for Checkstyle main/test sourcesets
   </description>
+
+  <rule ref="category/java/bestpractices.xml">
+    <!-- We do not care about this minor overhead, we are not Android application and we do not
+         want to change visibility of methods. Package-private visibility should be avoid almost
+         always. -->
+    <exclude name="AccessorMethodGeneration"/>
+    <!-- The PMD mistakes the AbstractViolationReporter::log() as a debug log. -->
+    <exclude name="GuardLogStatement"/>
+  </rule>
+  <rule ref="category/java/bestpractices.xml/AvoidPrintStackTrace">
+    <properties>
+      <!-- It is ok to use print stack trace in CLI class. -->
+      <property name="violationSuppressXPath" value="//ClassOrInterfaceDeclaration[@Image='Main']"/>
+    </properties>
+  </rule>
+  <rule ref="category/java/bestpractices.xml/JUnitAssertionsShouldIncludeMessage">
+    <properties>
+      <!-- The PMD check accepts only a string constant as the assert message.
+           This check uses a method call. -->
+      <property name="violationSuppressXPath" value="//ClassOrInterfaceDeclaration[@Image='JavadocCatchCheck']//MethodDeclaration[@Name='visitJavadocToken']"/>
+    </properties>
```

Comment:
Not sure about this. The code is
```
        public void visitJavadocToken(DetailNode ast) {
            assertEquals(ast.toString(), "JAVADOC", ast.getText());
            final DetailNode text = JavadocUtils.findFirstToken(ast, JavadocTokenTypes.TEXT);
            Assert.assertNotNull("Empty javadoc text at " + ast, text);
            assertEquals(ast.toString(), "Javadoc", text.getText());
            javadocsNumber++;
        }
```
`ast.toString()` is not a string, but this code works as it should. This seems to be a PMD bug.

Reply:
http://pmd.sourceforge.net/pmd-4.3.0/rules/junit.html#JUnitAssertionsShouldIncludeMessage
> JUnit assertions should include a message - i.e., use the three argument version of assertEquals(), not the two argument version. 

http://pmd.sourceforge.net/pmd-4.3.0/xref/net/sourceforge/pmd/rules/junit/JUnitAssertionsShouldIncludeMessage.html#45
> node.getArgumentCount() == args 

@pbludov Most likely this isn't a bug because PMD is just counting method arguments. You could file an issue with them for a feature to examine the first argument and not allow a only variable for it.

Reply:
@rnveach the code was changed since PMD 4.3. Now it is
https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/JUnitAssertionsShouldIncludeMessageRule.java#L69
```
        checks.add(new AssertionCall("assertEquals", 3) {
            @Override
            protected boolean isException(ASTArguments node) {
                List<ASTExpression> arguments = node.findDescendantsOfType(ASTExpression.class);
                boolean isExceptionJunit4 = isStringTypeOrNull(arguments.get(0));
                boolean isExceptionJunit5 = isStringTypeOrNull(arguments.get(2));

                return isExceptionJunit4 || isExceptionJunit5;
            }
        });
    }
 
   private boolean isStringTypeOrNull(ASTExpression node) {
        return node.getType() == String.class || node.getType() == null;
}
```
The rule wants a string constant now.

Reply:
Sorry, I didn't realize the version number in the link.

Reply:
@pbludov 
> https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/JUnitAssertionsShouldIncludeMessageRule.java#L69

The code was added just 12 days ago. It might not be released yet.